File: guides/micronaut-java-records/java/src/main/java/example/micronaut/BookRepository.java
package example.micronaut;

import io.micronaut.data.jdbc.annotation.JdbcRepository;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.repository.CrudRepository;
import java.util.List;

@JdbcRepository(dialect = Dialect.POSTGRES) // <1>
public interface BookRepository extends CrudRepository<Book, String> { // <2>
    List<BookCard> find(); // <3>
}

File: guides/micronaut-metrics/python/src/example/micronaut/book_repository.py
from typing import Annotated

from jakarta.validation.constraints import NotBlank
from java.util import Optional
from micronaut.data.jdbc.annotation import JdbcRepository
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.repository import CrudRepository

from .book import Book


@JdbcRepository(dialect=Dialect.H2)  # <1>
class BookRepository(CrudRepository[Book, int]):  # <2>
    def findByIsbn(self, isbn: Annotated[str, NotBlank]) -> Optional[Book]: ...
